keps/sig-node/1769-memory-manager/README.md

```diff
@@ -0,0 +1,782 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+The *Memory Manager* is a new component in *kubelet* ecosystem proposed to enable the feature of guaranteed memory (and hugepages) allocation for pods in Guaranteed QoS class. The feature offers a couple of allocation strategies. The first one, the *single-NUMA* strategy, is intended for high-performance and performance-sensitive applications. Relevant user stories are illustrated in here for the *single-NUMA* strategy. The second one, the *multi-NUMA* strategy, complements the overall design while it overcomes the situation that cannot be managed with the *single-NUMA* strategy. Namely, whenever the amount of memory demanded by a pod is in the excess of a single NUMA node capacity, the guaranteed memory is provisioned across multiple NUMA nodes with the *multi-NUMA* strategy.  
+
+In both scenarios, the *Memory Manager* employs hint generation protocol to yield the most suitable NUMA affinity for a pod, and it feeds the central manager (*Topology Manager*) with those affinity hints. Moreover, *Memory Manager* ensures that the memory which a pod requests is allocated from a minimum number of NUMA nodes.
+
+The up-to-date implementation of the *Memory Manager* is available through the link in [Implementation History](##implementation-history) section. Technically, the *single-NUMA* strategy is a special case of the *multi-NUMA* strategy, and thus, no separate implementation was developed for them. Still, the proposal differentiates between the *single-NUMA* and the *multi-NUMA* strategies. First of all, it allows for a number of important use cases or situations to be discussed and contrasted.       
+
+## Motivation
+
+### Goals
+
+- Offer guaranteed memory (and hugepages) allocation over a minimum number of NUMA nodes for containers (within a pod). 
+
+- Guaranteeing the affinity of memory and hugepages to the same NUMA node for the whole group of containers (within a pod). This is a long-term goal which will be achieved along with PR [#1752](https://github.com/kubernetes/enhancements/pull/1752) and the implementation of `hintprovider.GetPodLevelTopologyHints()` API in the *Memory Manager*.  
+
+### Non-Goals
+
+- Both the upgrade of K8S scheduler and further extension of the official pod specification are out of scope.
+
+- This proposal confines to the family of Linux operating systems.
+
+## Proposal
+
+The *Memory Manager* is a new component of _kubelet_ ecosystem proposed to enable single-NUMA and multi-NUMA guaranteed memory allocation. Besides:
+- *Memory Manager* is also a new hint provider for _Topology Manager_;
+- The hints are intended to indicate preferred memory (and hugepages) affinity which pins the memory for a container either to a single or a group of NUMA nodes;
+- By design, *Memory Manager* offers guaranteed memory allocation for pods in Guaranteed QoS class.
```

Comment:
In [the most recent issue (reported in 2020)](https://bugzilla.redhat.com/show_bug.cgi?id=1855067#c117) that you mentioned above, `cgroup.memory=nokmem` should be used as Andrea suggested to disable _"kernel memory accounting"_:

> The above was measured before the OOM kill, the dump was taken after, and they
> differ by 56kbytes which means all memory was tracked correctly and the bulk of
> it is in 32kbyte slabs (because the kernel wasn't booted with slub_max_order=0)
> for each CPU (and the only way to disable the per-memcg slab is to disable kmem
> by booting the kernel with "cgroup.memory=nokmem").

But the issue related to _"kernel memory accounting"_ popped up before (reported in 2018) and was discussed [[here]](https://github.com/kubernetes/kubernetes/issues/61937). 

In the discussion, [this recommendation](https://github.com/kubernetes/kubernetes/issues/61937#issuecomment-391057022) were provided by @dashpole on how to configure K8S when _"kernel memory accounting"_ is disabled:

> 1. Follow this comment to disable kernel memory accounting by recompiling the kernel: opencontainers/runc#1725 (comment).
> 2. I did some testing with kernel memory accounting disabled, and found that it made a relatively small impact on the ability of the kubelet to manage memory. I would recommend increasing --eviction-hard's memory.available parameter by 50Mi when disabling kernel memory accounting. 

> the way the kubelet manages memory is roughly to compare memory.usage_in_bytes to memory_capacity - eviction_threshold. Because memory.usage_in_bytes will not include kernel memory, the measured usage_in_bytes will be lower than the actual, and we may OOM even when usage_in_bytes is not close to 0. In my testing, I found that there was 30-50 Mi of remaining kernel memory when the node is under memory pressure. One way to compensate for this is to increase the eviction threshold.

@derekwaynecarr Would you expect us to provide similar recommendation for multi-NUMA use case? (as you already mentioned in another comment _"the multi-numa scenario increases the impact of cgroup.kmem behavior when sizing these workloads."_)

Reply:
I see it some serious cons to set `cgroup.memory=nokmem`, from the above bug:

- a malicious container can take over a ton of kernel memory, much more than the memory.max limit, potentially causing a DoS to other containers or system apps

- a malicious container can fragment forever the whole physical address space rendering all page blocks non-movable, so all other containers will forever run slow even after the malicious container has been destroyed

- for non-malicious containers, there's still a QoS concern, where one container doing lots of files I/O with zillions of tiny files could reduce the cache available to other containers

Also important to notice that kernel 5.9 has changed from having per memory cgroup kmemcahe to use a common kmemcache for all non-root memory cgroups. As a result, we should be seeing a 50% or more reduction in kernel slab memory usage in the containers.

So for the near future, I think it will be enough to add a section under container resources regarding the guaranteed pods and kmem consumption.

We can add such notice under our KEP, but IMHO it should be placed under the k8s documentation. @derekwaynecarr WDT?

Reply:
@cynepco3hahue I agree with above analysis, and a note about the more recent kernels is sufficient.